Source/Mothership/MPIHandlers.cpp

```diff
@@ -0,0 +1,497 @@
+
+    /* Pull message contents. */
+    std::string appName;
+    if (!decode_string_message(message, &appName))
+    {
+        debug_post(591, 3, "Q::CMND,Q::RUN", hex2str(message->Key()).c_str(),
+                   "Failed to decode.");
+        return 0;
+    }
+
+    debug_post(591, 3, "Q::CMND,Q::RUN", hex2str(message->Key()).c_str(),
+               dformat("appName=%s", appName.c_str()).c_str());
+
+    /* Get the application */
+    appInfo = appdb.check_create_app(appName);
+
+    /* Stage the run command, and run it now if possible. */
+    appInfo->stage_run();
+    if (appInfo->state == READY) run_application(appInfo);
+    return 0;
+}
+
+unsigned Mothership::handle_msg_cmnd_stop(PMsg_p* message)
+{
+    AppInfo* appInfo;
+
+    /* Pull message contents. */
+    std::string appName;
+    if (!decode_string_message(message, &appName))
+    {
+        debug_post(591, 3, "Q::CMND,Q::STOP", hex2str(message->Key()).c_str(),
+                   "Failed to decode.");
+        return 0;
+    }
+
+    debug_post(591, 3, "Q::CMND,Q::STOP", hex2str(message->Key()).c_str(),
+               dformat("appName=%s", appName.c_str()).c_str());
+
+    /* Get the application */
+    appInfo = appdb.check_create_app(appName);
+
+    /* Stage the stop command, and run it now if possible. */
+    appInfo->stage_stop();
+    if (appInfo->state == RUNNING) stop_application(appInfo);
+    return 0;
+}
+
+unsigned Mothership::handle_msg_bend_cnc(PMsg_p* message)
+{
+    uint8_t opcode;
+
+    /* Pull message contents. */
+    std::vector<P_Pkt_t> packets;
+    if (!decode_packets_message(message, &packets))
+    {
+        debug_post(591, 3, "Q::BEND,Q::CNC", hex2str(message->Key()).c_str(),
+                   "Failed to decode.");
+        return 0;
+    }
+
+    debug_post(591, 3, "Q::BEND,Q::CNC", hex2str(message->Key()).c_str(),
+               dformat("%lu packet(s).", packets.size()).c_str());
+
+    /* For each packet, get its opcode, and call the appropriate packet
+     * handling method, Posting if we don't recognise it. */
+    for (std::vector<P_Pkt_t>::iterator packetIt = packets.begin();
+         packetIt != packets.end(); packetIt++)
+    {
+        opcode = ((*packetIt).header.swAddr & P_SW_OPCODE_MASK)
+            >> P_SW_OPCODE_SHIFT;
+        switch (opcode)
+        {
+            case P_CNC_INSTR : handle_pkt_instr(&*packetIt); break;
+            case P_CNC_LOG : handle_pkt_log(&*packetIt); break;
+            case P_CNC_BARRIER : handle_pkt_barrier(&*packetIt); break;
+            case P_CNC_STOP : handle_pkt_stop(&*packetIt); break;
+            case P_CNC_KILL : handle_pkt_kill(&*packetIt); break;
+            default : Post(514, hex2str(opcode));
+        }
+    }
+
+    return 0;
+}
+
+unsigned Mothership::handle_msg_bend_supr(PMsg_p* message)
+{
+    int rc;
+
+    /* Get the application from the message. Note that we don't get the packets
+     * here because the supervisor SO reads the message (not the packets). */
```

Comment:
Why?

Reply:
I think I was thinking that passing the packets to user-defined supervisor would make sense, but then I realised that it doesn't matter because we generate the interface, so messages is fine